app/Http/Controllers/Widgets/DeviceLocationController.php

```diff
@@ -0,0 +1,48 @@
+<?php
+
+namespace App\Http\Controllers\Widgets;
+
+use Illuminate\Http\Request;
+use App\Http\Controllers\Controller;
+use App\Models\Device;
+use App\Models\Location;
+use LibreNMS\Config;
+use DB;
+#use Illuminate\Support\Facades\DB;
+
+class DeviceLocationController extends WidgetController
+{
+    protected $title = 'Device Location Summary';
+
+    #public function getSettingView(Request $request)
+    #{
+    #
+    #}
+
+    public function getView(Request $request)
+    {
+
+        $data = $this->getSettings();
+
+        $data['locations'] = DB::select("select 
```

Comment:
This should work

```php
$data['locations'] = Location::orderBy('location')->get()->map(function ($location) {
    /** @var Location $location */
    return [
        'location' => $location->location,
        'total' => $location->devices()->count(),
        'up' => $location->devices()->isUp()->count(),
        'down' => $location->devices()->isDown()->count(),
        'ignored' => $location->devices()->isIgnored()->count(),
        'disabled' => $location->devices()->isDisabled()->count(),
    ];
});
```

I'm not sure the totals belongs in this widget... there are other ones that do that.

Basically you can use the same queries as above for the totals if you want them.
I'm also not sure why you have included the id.

Reply:
@murrant That worked thanks.

How do I add a row to the widget table?      
Need to add the following to the widget table. But i think it's seeded from somewhere else.
widget_id=17 | Device Location           | device-location

Reply:
@murrant bump.... looking to know how to add a row to the widgets table.

Reply:
@esundberg you will need to create a [database migration](https://laravel.com/docs/migrations) and update the seeder database/seeds/DefaultWidgetSeeder.php

Reply:
Ok thanks I will get to this shortly

Reply:
I ran into an issue while running this in our test environment. Need to make some correction. It's displaying locations that have no devices associated to it.

Reply:
@murrant It looks like the location's are not being removed and always being added. How would if make the query so it only returns the location if there a 1 or more devices at that location.

Location::orderBy('location') WHERE count > 0

Reply:
Location->whereHas('device')->orderBy('location')->get();

Reply:
Ok thanks. i will get this push through when i have some time.